xilem_web/src/concurrent/await_once.rs

```diff
@@ -0,0 +1,118 @@
+///
+/// # Examples
+///
+/// ```
+/// use xilem_web::{core::fork, concurrent::await_once, elements::html::div, interfaces::Element};
+///
+/// fn app_logic(state: &mut i32) -> impl Element<i32> {
+///     fork(
+///         div(*state),
+///         await_once(
+///             |_state: &mut i32| std::future::ready(42),
+///             |state: &mut i32, meaning_of_life| *state = meaning_of_life,
+///         )
+///     )
+/// }
+/// ```
+pub fn await_once<State, Action, FOut, F, InitFuture, OA, Callback>(
+    init_future: InitFuture,
+    callback: Callback,
+) -> AwaitOnce<InitFuture, Callback, State, Action>
+where
+    State: 'static,
+    Action: 'static,
+    FOut: std::fmt::Debug + 'static,
+    F: Future<Output = FOut> + 'static,
+    InitFuture: Fn(&mut State) -> F + 'static,
+    OA: OptionalAction<Action> + 'static,
+    Callback: Fn(&mut State, FOut) -> OA + 'static,
+{
+    AwaitOnce {
+        init_future,
+        callback,
+        phantom: PhantomData,
+    }
+}
+
+pub struct AwaitOnceState<F> {
+    future: Option<F>,
+}
+
+impl<State, Action, InitFuture, F, FOut, Callback, OA> View<State, Action, ViewCtx, DynMessage>
+    for AwaitOnce<InitFuture, Callback, State, Action>
+where
+    State: 'static,
+    Action: 'static,
+    FOut: std::fmt::Debug + 'static,
+    F: Future<Output = FOut> + 'static,
+    InitFuture: Fn(&mut State) -> F + 'static,
+    OA: OptionalAction<Action> + 'static,
+    Callback: Fn(&mut State, FOut) -> OA + 'static,
+{
+    type Element = NoElement;
+
+    type ViewState = AwaitOnceState<F>;
+
+    fn build(&self, ctx: &mut ViewCtx) -> (Self::Element, Self::ViewState) {
+        let thunk = ctx.message_thunk();
+        // we can't directly push the initial message, as this would be executed directly (not in the next microtick), which in turn means that the already mutably borrowed `App` would be borrowed again.
+        // So we have to delay this with a spawn_local
+        spawn_local(async move { thunk.push_message(None::<FOut>) });
+        (NoElement, AwaitOnceState { future: None })
+    }
+
+    fn rebuild<'el>(
+        &self,
+        _prev: &Self,
+        view_state: &mut Self::ViewState,
+        ctx: &mut ViewCtx,
+        (): Mut<'el, Self::Element>,
+    ) -> Mut<'el, Self::Element> {
+        if let Some(future) = view_state.future.take() {
+            let thunk = ctx.message_thunk();
+            spawn_local(async move { thunk.push_message(Some(future.await)) });
+        }
+    }
+
+    fn teardown(&self, _: &mut Self::ViewState, _: &mut ViewCtx, _: Mut<'_, Self::Element>) {}
+
+    fn message(
+        &self,
+        view_state: &mut Self::ViewState,
+        id_path: &[ViewId],
+        message: DynMessage,
+        app_state: &mut State,
+    ) -> MessageResult<Action, DynMessage> {
+        assert!(id_path.is_empty()); // `debug_assert!` instead? to save some bytes in the release binary?
```

Comment:
..yes, I almost always use `debug_assert`, but mostly in projects where the potential bugs should actually be noticed during development. I don't know enough about the internals of Xilem to judge the likelihood of this.
In any case, it is an advantage to create a panic as early as possible if something is obviously fundamentally wrong.
Since Xilem is still under development, an early bug report from users is probably also very desirable.
In other words: at the moment, it would be even more important to get early feedback on errors than to save bytes :wink:

Reply:
Yes, my thinkings similar, it's probably better to get bug reports when this is somehow triggered.